tests/frontends/test_pytorch_ops.py

```diff
@@ -0,0 +1,458 @@
+
+    backend_options_dict = {
+        'pytorch': {'device': device_type},
+        'nnvm': {'target': device_type, 'device_id': 0},
+        'relay': {'target': device_type, 'device_id': 0}
+    }
+
+    backend_options = backend_options_dict[backend]
+
+    return backend_options
+
+
+# TODO: add relay support
+# TODO: maybe fixture for return_backend=True and return_backend=False
+@pytest.fixture(params=[
+    pytest.param('pytorch'),
+    pytest.param('nnvm')
+])
+def _backend_fixture(request):
+    return request.param
+
+
+class Args():
+
+    def __init__(self):
+        # device used
+        self.dev = 'cpu'
+        # backend used
+        self.backend = 'pytorch'
+        # numerical precision
+        self.dtype = 'float32'
+
+
+args = Args()
+
+
+def compare_fwd(*tests, optimize=True, python=True, profile=False):
+    """Decorate a function to parse and run it against pure Python.
+
+    Returns a unit test that will parse the function, and then for
+    each `inputs` tuple in `tests` it will check that the pure Python,
+    undecorated function returns that same output.
+
+    This uses the full myia pipeline.
+
+    Arguments:
+        tests: One or more inputs tuple.
+
+    """
+    fwd_pipeline = compile_pipeline if optimize else \
+        compile_pipeline.configure({'opt.phases.main': []})
+
+    def decorate(fn):
+        def test(args):
+            nonlocal profile
+            if not isinstance(args, tuple):
+                args = (args,)
+
+            _fwd_test(fn, args,
+                      pipeline=fwd_pipeline,
+                      optimize=optimize,
+                      python=python,
+                      profile=profile)
+
+        m = mark.parametrize('args', list(tests))(test)
+        m.__orig__ = fn
+        return m
+    return decorate
+
+
+grad_whitelist = whitelist | {P.J, P.Jinv}
+
+
+@validate_abstract.variant
+def grad_validate_abstract(self, t: AbstractJTagged):
+    pass
+
+
+step_grad_validate = Validator.partial(
+    whitelist=grad_whitelist,
+    validate_abstract=grad_validate_abstract
+)
+
+
+@pipeline_function
+def grad_wrap(self, graph):
+    if isinstance(graph, Primitive):
+        jg = realJ(graph, self.resources)
+        g = grad.make_gf(jg, jg.parameters, wrt=range(len(jg.parameters)),
+                         dbg=jg.debug, sens_param=True)
+    else:
+        g = grad.make_gf(graph, graph.parameters,
+                         wrt=range(len(graph.parameters)),
+                         dbg=graph.debug, sens_param=True,
+                         apply_j=True)
+    return g
+
+
+grad_pipeline = PipelineDefinition(
+    resources=standard_resources,
+    steps=dict(
+        parse=steps.step_parse,
+        resolve=steps.step_resolve,
+        infer=steps.step_infer,
+        specialize=steps.step_specialize,
+        opt=steps.step_debug_opt,
+        validate=step_grad_validate,
+        # compile=steps.step_compile,
+        export=steps.step_debug_export,
+    )
+)
```

Comment:
You should import `grad_whitelist`, `grad_validate_abstract`, `grad_wrap` and `grad_pipeline` from `test_grad`, to avoid unnecessary code duplication.